### PR TITLE
Feat (brevitas_examples/llm): BOS preprocessing for calibration data

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR
Follow up to #1229 

## Changes Made in this PR

The final results of the discussion is to add the possibility to pre-prepocess the calibration datasets, adding BOS token whenever needed, before fusing and splitting the datasets into chunks of seqlength length.

We also preserve the possibility to not do this, and thus have a single BOS token at the beginning of the entire calibration set, which doesn't make a whole lot of sense logically but it's a widespread setup

## Testing Summary

Check issue above for results

## Checklist

- [ ] Code comments added to any hard-to-understand areas, if applicable.
- [ ] Changes generate no new warnings.
- [ ] Updated any relevant tests, if applicable.
- [ ] No conflicts with destination `dev` branch.
- [ ] I reviewed my own code changes.
- [ ] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

**Comment by Giuseppe5** [comment]
Re-posting here from @nickfraser 

> 
> Sanity-checking results with this change - I'm seeing for several Llama-3.x models @ FP4, that perplexity doesn't change much, but it's not always an improvement. See results below:
> 
>  model | config | float_ppl | quant_ppl(dev) | quant_ppl(PR) |
> | --- | --- | --- | --- | --- |
> | meta-llama/Llama-3.2-1B | 1 | 8.78 | 10.20 | 10.19 |
> | meta-llama/Llama-3.2-1B | 2 | 8.78 | 10.79 | 10.35 |
> | meta-llama/Llama-3.2-1B | 3 | 8.78 | 10.28 | 10.43 |
> | meta-llama/Llama-3.2-1B | 4 | 8.78 | 10.94 | 10.66 |
> | meta-llama/Llama-3.2-1B | 5 | 8.78 | 10.53 | 10.45 |
> | meta-llama/Llama-3.2-1B | 6 | 8.78 | 11.06 | 10.70 |
> | meta-llama/Llama-3.2-1B | 7 | 8.78 | 10.30 | 10.40 |
> | meta-llama/Llama-3.2-1B | 8 | 8.78 | 10.89 | 10.70 |
> | meta-llama/Llama-3.2-3B | 9 | 6.99 | 8.09 | 7.76 |
> | meta-llama/Llama-3.2-3B | 10 | 6.99 | 8.32 | 7.90 |
> | meta-llama/Llama-3.2-3B | 11 | 6.99 | 8.20 | 7.89 |
> | meta-llama/Llama-3.2-3B | 12 | 6.99 | 8.29 | 7.85 |
> 
> More results to come...

File: src/brevitas_examples/llm/llm_args.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def create_llm_args_parser():
     parser.add_argument(
         '--dataset',
         type=str,
-        choices=['wikitext2', 'c4'],
+        choices=['wikitext2', 'c4', 'wikitext2_clm', 'c4_clm', 'pile_clm'],
```
Comment (nickfraser) [inline src/brevitas_examples/llm/llm_args.py:48]: How does the `_clm` differentiate from the other versions? Looks like the preprocessing is handled by the `--no-bos-preprocessing` argument.

Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_args.py:48]: This is a mistake, I'll remove this

**Review by nickfraser — APPROVED** [review]
LGTM! Does this produce different values in our accuracy tests? These might need to be updated too in the full test suite.

File: src/brevitas_examples/llm/llm_quant/data.py [hunk 2]
```diff
@@ -63,54 +83,134 @@ def get_c4(
...
-                # Add BOS token.
-                if tokenizer.eos_token_id is not None:
-                    input_ids[:, 0] = tokenizer.eos_token_id
+    return dataset
 
-                dataset.append({"input_ids": input_ids, "attention_mask": attention_mask})
-                pbar.update(1)
 
-    return dataset
+def group_texts(examples: Dict[str, List[np.ndarray]],
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data.py:90]: Add comment indicating that the code is adapted from this [file](https://github.com/huggingface/nanotron/blob/main/src/nanotron/data/processing.py).

File: src/brevitas_examples/llm/llm_quant/data.py [hunk 3]
```diff
@@ -63,54 +83,134 @@ def get_c4(
...
+            _tokenize_and_group_texts,
+            tokenizer=tokenizer,
+            sequence_length=seqlen,
+            filter_empty_sequences=filter_empty_sequences),
+        input_columns=text_column_name,
+        remove_columns=data.column_names,
+        features=Features({"input_ids": Sequence(feature=Value(dtype="int64"), length=seqlen)}),
+        batched=True,
+        num_proc=dataset_processing_num_proc_per_process,
+        load_from_cache_file=False,
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/data.py:174]: `load_from_cache_file` was disabled while experimenting, but ideally it should be `True` for efficiency. As in nanotron,  I would a flag `dataset_overwrite_cache`  to `get_dataset_clm`, and then do `load_from_cache_file = not dataset_overwrite_cache`.

Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_quant/data.py:174]: Not sure if I want to add another flag. I'm going to hard-code it to True after the tests pass. If someone wants to overwrite their dataset, they might have to delete them manually

File: src/brevitas_examples/llm/llm_args.py [hunk 4]
```diff
@@ -45,7 +45,7 @@ def create_llm_args_parser():
     parser.add_argument(
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_args.py:45]: Do you think it is worth having separate arguments for passing the calibration/validation dataset? Almost always, we want to evaluate on wikitext2 for comparison with other works, which forces to set `---dataset wikitext2`, even though any other calibration dataset could be used.

File: src/brevitas_examples/llm/main.py [hunk 5]
```diff
@@ -543,7 +542,8 @@ def quantize_llm(args, extra_args=None):
             with torch.no_grad(), quant_inference_mode(model, compile=args.compile_eval):
                 model(**calibration_loader[0])
 
-                wrapped_model = HFLM(pretrained=model)  # need to wrap for LLM eval
+                wrapped_model = HFLM(
```
Comment (pablomlago) [inline src/brevitas_examples/llm/main.py:545]: I think `add_bos_token` is not currently exposed in `llm_args.py`. Maybe we should leave it as it was before, since the increment in performance was not very noticeable.

Comment (Giuseppe5) [reply src/brevitas_examples/llm/main.py:545]: It got misshandled during review, my bad.

Anyway, I don't have a strong opinion about whether to have a separate flag for this or just stick with the old behaviour. @nickfraser what do you think?

File: src/brevitas_examples/llm/llm_args.py (hunk 4 above)
Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_args.py:45]: Let's keep that for another PR
